feat: add asset existence check before replacing static URLs

bra-i-am opened the pull request:
## Description

This merge request solves an issue where the save warning is shown without making changes to new courses. Additionally, it also adds an existence check before replacing the static URLs

## Supporting information

https://github.com/openedx/frontend-app-authoring/issues/1265

## Testing instructions

1. Mount this MFE in a Tutor environment and git switch to this branch
2. Enable `contentstore.new_studio_mfe.use_new_schedule_details_page`
3. Create course in Studio
4. Open Settings -> Schedule & details
5. Once you enter this page, you shouldn't watch the warning popover to jump out as you just made a change
6. Additionally, in the `Course overview` it must be displayed the default images if the assets are not found

bra-i-am committed:
feat: add asset existence check before replacing static URLs in TinyMCE content

* fix: update image replacement in the initial value so the change is not taken as a user action

openedx-webhooks commented:
Thanks for the pull request, @bra-i-am!


This repository is currently maintained by `@openedx/2u-tnl`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

bra-i-am committed:
feat: add option to validate asset URLs before replacement in TinyMCE content

bra-i-am committed:
test: add unit test for updateCourseDetailsOverview in scheduleAndDetails slice

arbrandes reviewed (changes requested):
I think this is a reasonable approach to fix #1265.  Thank you!  I would like for it to use react-query, though, if at all possible.

bra-i-am committed:
feat: refactor course details management by removing deprecated actions and integrating new API hooks

bra-i-am committed:
fix: remove unnecessary error check for courseDetails in ScheduleAndDetails component

bra-i-am committed:
fix: remove unnecessary blank line in TinyMceEditor hooks test

bra-i-am committed:
fix: remove unused dispatch from TinyMceWidget component

bra-i-am commented:
@arbrandes, I have already addressed your comment and modified the code to implement the fix using react-query.

I'm looking forward to your feedback! Thanks :sparkles:

arbrandes commented:
Thanks, @bra-i-am!  I'm putting this back on my queue.  I'll try to get to it before the conference, but it's possible I'll only be able to review or on after July 7th.

mphilbrick211 commented:
Hi @bra-i-am! Is this pull request still in progress?

bra-i-am committed:
feat: add validateAssetUrl prop to TextEditor component

arbrandes commented on src/schedule-and-details/hooks.jsx:
```diff
@@ -54,7 +52,7 @@ const useSaveValuesPrompt = (
       setEditedValues(initialEditedData);
     }
-  }, [initialEditedData]);
+  }, [initialEditedData.isLoading]);
```
This means that the data is only set on initial load, not after saving.

I suggest you expose `dataUpdatedAt` (a react query primitive) from the query hook, then depend on that here instead of `isLoading`.

bra-i-am committed:
refactor: address feedback

bra-i-am committed:
fix: solve rebase issues

bra-i-am committed:
refactor: add custom hook to manage static replacement

mphilbrick211 commented:
Flagging this for you, @arbrandes

arbrandes reviewed (changes requested):
Hi @bra-i-am, thanks for sticking with it!  I unleashed Claude on the PR, and it seems there are a few more things to work on:

#### 1. React-query metadata leaks into the save API payload (bug)

[`useGetCourseDetails`](https://github.com/openedx/frontend-app-authoring/blob/fdbadd3abdf9f65200716283523b8ee77ff51387/src/schedule-and-details/data/apiHooks.ts#L7-L24) returns `{ ...data, id, isLoading, isError, refetch, dataUpdatedAt }`. This entire object is passed as `initialEditedData` to [`useSaveValuesPrompt`](https://github.com/openedx/frontend-app-authoring/blob/fdbadd3abdf9f65200716283523b8ee77ff51387/src/schedule-and-details/index.jsx#L98-L103), which stores it in `editedValues` state ([hooks.jsx#L43](https://github.com/openedx/frontend-app-authoring/blob/fdbadd3abdf9f65200716283523b8ee77ff51387/src/schedule-and-details/hooks.jsx#L43), [hooks.jsx#L53](https://github.com/openedx/frontend-app-authoring/blob/fdbadd3abdf9f65200716283523b8ee77ff51387/src/schedule-and-details/hooks.jsx#L53)).

When the user saves, `editedValues` is sent to the PUT API endpoint via `updateWithDefaultValues(editedValues)` -> `updateCourseDetails(courseId, details)` -> `convertObjectToSnakeCase(details)`. This means fields like `is_loading`, `is_error`, and `data_updated_at` are included in the request body. `refetch` (a function) would be dropped by JSON serialization, but the other fields would not.

Depending on the server's behavior, this could cause validation errors or silently pass unexpected data. At minimum, these fields should be stripped before `editedValues` is initialized.

A clean fix would be to separate the course details data from the react-query metadata in `useGetCourseDetails`, or to extract only the data fields when passing to `useSaveValuesPrompt`.

#### 2. 403 handling regression for course details

Previously, a 403 response when fetching course details triggered a dedicated "denied" state (`RequestStatus.DENIED`) that rendered a `<Placeholder />` component. The [base version of index.jsx](https://github.com/openedx/frontend-app-authoring/blob/abcef4a50269bbfc751a9578f476941b016874db/src/schedule-and-details/index.jsx#L85) checked `loadingDetailsStatus === RequestStatus.DENIED`.

Now, `useGetCourseDetails` only exposes `isError` (true for any error - 403, 404, 500, etc.), and the [denied check](https://github.com/openedx/frontend-app-authoring/blob/fdbadd3abdf9f65200716283523b8ee77ff51387/src/schedule-and-details/index.jsx#L148) only covers `loadingSettingsStatus`. A 403 on course details will now show a generic "load failed" alert instead of the placeholder. This is a behavior regression.

To preserve the existing behavior, the react-query hook could check `error.response.status === 403` and expose it separately, or react-query's `meta`/`onError` could be used to handle it.

#### 3. Asset validation uses GET instead of HEAD

[`replaceStaticWithAsset`](https://github.com/openedx/frontend-app-authoring/blob/fdbadd3abdf9f65200716283523b8ee77ff51387/src/editors/sharedComponents/TinyMceWidget/hooks.ts#L161-L169) calls `getAuthenticatedHttpClient().get(staticFullUrl)` to check whether an asset exists. For assets that could be large (images, PDFs), this downloads the entire file just to determine existence. The asset-serving view in edx-platform (`contentserver/views.py`) uses Django's `@require_safe` decorator, which supports HEAD requests. Using `.head(staticFullUrl)` instead of `.get()` would avoid transferring the response body.

#### 4. Whitespace: tab character in apiHooks.ts

[`apiHooks.ts` line 22](https://github.com/openedx/frontend-app-authoring/blob/fdbadd3abdf9f65200716283523b8ee77ff51387/src/schedule-and-details/data/apiHooks.ts#L22) uses a tab for `dataUpdatedAt` indentation instead of spaces, inconsistent with the rest of the file.

gabrieldamours commented:
Hi @bra-i-am! I'm just checking in on this PR as part of our [Spring Cleaning Hackathon](https://discuss.openedx.org/t/contributions-review-spring-cleaning-hackathon-starting-may-11/18919/2) — we're trying to clear PRs that have been waiting a while. It looks like a few changes are required here, but can you spare some time to push it to the finish line? Thanks :)

mphilbrick211 commented:
> Hi @bra-i-am! I'm just checking in on this PR as part of our [Spring Cleaning Hackathon](https://discuss.openedx.org/t/contributions-review-spring-cleaning-hackathon-starting-may-11/18919/2) — we're trying to clear PRs that have been waiting a while. It looks like a few changes are required here, but can you spare some time to push it to the finish line? Thanks :)

Friendly ping on this, @bra-i-am

dcoa commented:
Hi @bra-i-am , @arbrandes, @mphilbrick211, @bradenmacdonald 

I was checking this PR as part of the Hackathon and seems that the issue were reported twice and a solution for this was already merged. Taking that in consideration we can close this PR and the related issue. 

This is the current behavior of the platform (that is the expected one)

[Screencast from 2026-06-10 17-33-15.webm](https://github.com/user-attachments/assets/971cd6f8-9cf9-4931-a8b4-43766e2165b6)


### Extra information

- Issue related to the current PR that can be closed https://github.com/openedx/frontend-app-authoring/issues/1265
- Current Implementation PR https://github.com/openedx/frontend-app-authoring/pull/2702  and related issue https://github.com/openedx/frontend-app-authoring/issues/2668


If everyone agree and there is no new input in the coming days I will proceed to close it on Monday.

dcoa commented:
I am closing this PR given the explanation above. Feel free to reopen if necessary.
